tiktoken-rs/src/api.rs

```diff
@@ -66,6 +47,39 @@ pub fn get_completion_max_tokens(model: &str, prompt: &str) -> Result<usize> {
     Ok(context_size.saturating_sub(prompt_tokens))
 }
 
+/// Calculates the number of tokens for chat completion based on the model and messages provided.
+/// Based on https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
+pub fn num_tokens_from_messages(model: &str, messages: &[ChatCompletionRequestMessage]) -> Result<usize> {
+    let tokenizer =
+        get_tokenizer(model).ok_or_else(|| anyhow!("No tokenizer found for model {}", model))?;
+    if tokenizer != Tokenizer::Cl100kBase {
+        bail!("Chat completion is only supported chat models")
+    }
+    let bpe = get_bpe_from_tokenizer(tokenizer)?;
+
+    let (tokens_per_message, tokens_per_name) = if model.starts_with("gpt-3.5") {
+        (
+            4, // every message follows <im_start>{role/name}\n{content}<im_end>\n
+            -1 // if there's a name, the role is omitted
+        )
+    } else {
+        (3, 1)
+    };
+
+    let mut num_tokens: i32 = 0;
+    for message in messages {
+        num_tokens += tokens_per_message;
+        num_tokens += bpe.encode_with_special_tokens(&message.role.to_string()).len() as i32;
+        num_tokens += bpe.encode_with_special_tokens(&message.content).len() as i32;
+        if let Some(name) = &message.name {
+            num_tokens += bpe.encode_with_special_tokens(name).len() as i32;
```

Comment:
Keep everything as `usize`. We shouldn't have negative values here, and usize is best for portability